graph/layout/eades_test.go

```diff
@@ -0,0 +1,205 @@
+			6: r2.Vec{X: 14.236725269213267, Y: 14.76680859469247},
+			7: r2.Vec{X: 15.843133632993109, Y: 14.895887859745487},
+			8: r2.Vec{X: 17.14673527193477, Y: 14.779176044572363},
+		},
+	},
+	{
+		name: "tube",
+		g: func() graph.Graph {
+			edges := []simple.Edge{
+				{simple.Node(0), simple.Node(1)},
+				{simple.Node(0), simple.Node(2)},
+				{simple.Node(0), simple.Node(3)},
+				{simple.Node(1), simple.Node(2)},
+				{simple.Node(1), simple.Node(4)},
+				{simple.Node(2), simple.Node(5)},
+				{simple.Node(3), simple.Node(4)},
+				{simple.Node(3), simple.Node(5)},
+				{simple.Node(3), simple.Node(6)},
+				{simple.Node(4), simple.Node(5)},
+				{simple.Node(4), simple.Node(7)},
+				{simple.Node(5), simple.Node(8)},
+				{simple.Node(6), simple.Node(7)},
+				{simple.Node(6), simple.Node(8)},
+				{simple.Node(7), simple.Node(8)},
+			}
+			g := simple.NewUndirectedGraph()
+			for _, e := range edges {
+				g.SetEdge(e)
+			}
+			return orderedGraph{g}
+		}(),
+		param: Eades{C1: 2, C2: 1, C3: 1, C4: 0.1, M: 100, Theta: 0.1, Src: rand.NewSource(1)},
+		want: map[int64]r2.Vec{
+			0: r2.Vec{X: 20.81592592580858, Y: 7.093353250039021},
+			1: r2.Vec{X: 21.726993818602256, Y: 7.055585705916057},
+			2: r2.Vec{X: 21.810995821173716, Y: 5.783565022523372},
+			3: r2.Vec{X: 22.0864584730445, Y: 8.300612169610107},
+			4: r2.Vec{X: 23.00743865904811, Y: 8.184144796089953},
+			5: r2.Vec{X: 23.20316516390141, Y: 6.828096609467102},
+			6: r2.Vec{X: 23.007775564315455, Y: 9.319231864784014},
+			7: r2.Vec{X: 23.97755074521381, Y: 9.266038395576885},
+			8: r2.Vec{X: 24.15211052582012, Y: 7.809512061906321},
+		},
+	},
+}
+
+func TestEades(t *testing.T) {
+	// Factor to dilate positions so graphviz separates nodes.
+	const scaling = 100
+
+	for _, test := range eadesTests {
+		eades := test.param
+		o := NewOptimizer(test.g, eades.Update)
+		var n int
+		for o.Update() {
+			n++
+		}
+		if n != test.param.M {
+			t.Errorf("unexpected number of iterations: got:%d want:%d", n, test.param.M)
+		}
+		got := make(map[int64]r2.Vec)
+		nodes := test.g.Nodes()
+		for nodes.Next() {
+			id := nodes.Node().ID()
+			got[id] = o.Location(id)
+		}
+		if !reflect.DeepEqual(got, test.want) {
+			t.Errorf("unexpected result:\ngot: %#v\nwant:%#v", got, test.want)
+		}
+
+		// TODO(kortschak): Replace this with Go rendering code.
+		if _, err := exec.LookPath("neato"); err != nil {
+			continue
+		}
+		layout := simple.NewUndirectedGraph()
+		for _, u := range graph.NodesOf(test.g.Nodes()) {
+			for _, v := range graph.NodesOf(test.g.From(u.ID())) {
+				if v.ID() < u.ID() {
+					continue
+				}
+				layout.SetEdge(simple.Edge{
+					F: positionNode{id: u.ID(), pos: got[u.ID()].Scale(scaling)},
+					T: positionNode{id: v.ID(), pos: got[v.ID()].Scale(scaling)},
+				})
+			}
+		}
+		b, _ := dot.Marshal(layout, test.name, "", "  ")
+		path := filepath.Join("testdata", test.name+".png")
+		cmd := exec.Command("neato", "-n", "-Tpng", "-o", path)
```

Comment:
perhaps document somehow (and somewhere) that `neato` comes with graphviz?
(and perhaps also test for that command and disable the test if it's not available?)

Reply:
This is already disabled if `neato` is not in `$PATH`. It's only temporary until I implement a plotter for graphs anyway.